[UnifiedPDF] Do more targeted repaints when annotations change

smfr opened the pull request:
#### 28ce0bd231af1910f25843f7a6cbed856c1aaf78
<pre>
[UnifiedPDF] Do more targeted repaints when annotations change
<a href="https://bugs.webkit.org/show_bug.cgi?id=269546">https://bugs.webkit.org/show_bug.cgi?id=269546</a>
<a href="https://rdar.apple.com/123066080">rdar://123066080</a>

Reviewed by Sammy Gill.

Replace the call to updateLayerHierarchy() for annotation changes to more focused repaints.

First, we have `m_annotationTrackingState.startAnnotationTracking`/`finishAnnotationTracking`
go through UnifiedPDFPlugin wrappers, which can do the repaint. Then we change those
functions to return a value to indicate if a repaint is needed. This repaint needs to also
throw away cached tiles in some cases, so do it via `setPDFChangedInDocumentRect()`.

Add `documentRectForAnnotation()`. This rect is then used to do the repaints.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::invalidateTilesForPaintingRect):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPDFChangedInDocumentRect):
(WebKit::UnifiedPDFPlugin::paintHoveredAnnotationOnPage):
(WebKit::UnifiedPDFPlugin::pageIndexWithHoveredAnnotation const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::documentRectForAnnotation const):
(WebKit::UnifiedPDFPlugin::startTrackingAnnotation):
(WebKit::UnifiedPDFPlugin::finishTrackingAnnotation):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):
(WebKit::AnnotationTrackingState::handleMouseDraggedOffTrackedAnnotation):

Canonical link: <a href="https://commits.webkit.org/274886@main">https://commits.webkit.org/274886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da3d1b1591007f79a4b8c90052f1bd0b526ff0b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42777 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16573 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16204 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34731 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44055 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36485 "1 api test failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36013 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39762 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12337 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16666 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9040 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -54,15 +54,20 @@ enum class WebEventType : uint8_t;
 enum class WebMouseEventButton : int8_t;
 enum class WebEventModifier : uint8_t;
 
+enum class RepaintOrNot : bool { No, Yes };
```
How about `ShouldRepaint`? Seems more in line with our usual naming I think.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -351,8 +356,13 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     std::optional<PDFDocumentLayout::PageIndex> pageIndexWithHoveredAnnotation() const;
     void paintHoveredAnnotationOnPage(PDFDocumentLayout::PageIndex, WebCore::GraphicsContext&, const WebCore::FloatRect& clipRect);
 
+    WebCore::FloatRect contentsRectForAnnotation(RetainPtr<PDFAnnotation>) const;
```
Why is this passed by value?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
         scrollToPDFDestination(destination);
 }
 
+WebCore::FloatRect UnifiedPDFPlugin::contentsRectForAnnotation(RetainPtr<PDFAnnotation> annotation) const
```
Why is this taking a `RetainPtr<>`? The implementation doesn't `WTFMove()` it so there is no benefit. I think this should be a raw pointer.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+
+    if (repaint == RepaintOrNot::Yes) {
+        auto annotationRepaintRect = contentsRectForAnnotation(m_annotationTrackingState.trackedAnnotation());
+        m_contentsLayer->setNeedsDisplayInRect(annotationRepaintRect);
```
Need a RefPtr on the stack for m_contentsLayer

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    auto repaint = m_annotationTrackingState.finishAnnotationTracking(mouseEventType, mouseEventButton);
+
+    if (repaint == RepaintOrNot::Yes)
+        m_contentsLayer->setNeedsDisplayInRect(annotationRepaintRect);
```
Needs a RefPtr on the stack for m_contentsLayer

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+{
+    auto repaint = m_annotationTrackingState.startAnnotationTracking(WTFMove(annotation), mouseEventType, mouseEventButton);
+
+    if (repaint == RepaintOrNot::Yes) {
```
Could move this call into the if

smfr replied:
I find this more readable.

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    auto repaint = m_annotationTrackingState.startAnnotationTracking(WTFMove(annotation), mouseEventType, mouseEventButton);
+
+    if (repaint == RepaintOrNot::Yes) {
+        auto annotationRepaintRect = contentsRectForAnnotation(m_annotationTrackingState.trackedAnnotation());
```
Do we actually need to store this?

smfr replied:
I find this more readable.

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1717,6 +1710,43 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    }
+}
+
+void UnifiedPDFPlugin::finishTrackingAnnotation(WebEventType mouseEventType, WebMouseEventButton mouseEventButton)
```
Same comments as in `startTrackingAnnotation`

sammygill reviewed (approved):
Annotation changes look good, agree with Chris' smart pointer suggestions

smfr committed:
[UnifiedPDF] Do more targeted repaints when annotations change
https://bugs.webkit.org/show_bug.cgi?id=269546
rdar://123066080

Reviewed by Sammy Gill.

Replace the call to updateLayerHierarchy() for annotation changes to more focused repaints.

First, we have `m_annotationTrackingState.startAnnotationTracking`/`finishAnnotationTracking`
go through UnifiedPDFPlugin wrappers, which can do the repaint. Then we change those
functions to return a value to indicate if a repaint is needed. This repaint needs to also
throw away cached tiles in some cases, so do it via `setPDFChangedInDocumentRect()`.

Add `documentRectForAnnotation()`. This rect is then used to do the repaints.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::invalidateTilesForPaintingRect):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPDFChangedInDocumentRect):
(WebKit::UnifiedPDFPlugin::paintHoveredAnnotationOnPage):
(WebKit::UnifiedPDFPlugin::pageIndexWithHoveredAnnotation const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::documentRectForAnnotation const):
(WebKit::UnifiedPDFPlugin::startTrackingAnnotation):
(WebKit::UnifiedPDFPlugin::finishTrackingAnnotation):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):
(WebKit::AnnotationTrackingState::handleMouseDraggedOffTrackedAnnotation):

Canonical link: https://commits.webkit.org/274886@main
